lib/linter/linter.js

```diff
@@ -942,7 +942,9 @@ function runRules(sourceCode, configuredRules, ruleMapper, parserOptions, parser
     });
 
     // only run code path analyzer if the top level node is "Program", skip otherwise
-    const eventGenerator = nodeQueue[0].node.type === "Program" ? new CodePathAnalyzer(new NodeEventGenerator(emitter)) : new NodeEventGenerator(emitter);
+    const eventGenerator = nodeQueue[0].node.type === "Program"
+        ? new CodePathAnalyzer(new NodeEventGenerator(emitter, { visitorKeys: sourceCode.visitorKeys, fallback: Traverser.getKeys }))
+        : new NodeEventGenerator(emitter, { visitorKeys: sourceCode.visitorKeys, fallback: Traverser.getKeys });
```

Comment:
Thank you for letting me know. I wasn't aware that it would work without any options.
I changed it to use `Object.keys()` when it encountered an unknown node in the PR of esquery, so in some cases the option didn't seem to be needed.

However, if we omit the option, it cannot traverse unknown keys. I have added a test case to check this.
https://github.com/eslint/eslint/pull/14072/files#diff-0e70eee0ed29d60a149f39870d312c97bc4f6346a0f2508196df518ebf3f1602R381

This test case checks that `B` matches `:first-child` in the following TS source code.

```ts
class A implements B {}
```

Reply:
> Can we add some linter tests that would fail without this change?

sorry. I missed this. I will add test cases from now on.

Reply:
I added test case.

Reply:
Great, thanks! The test indeed fails if `Linter` doesn't pass visitor keys to `NodeEventGenerator`.

Nice idea to use known nodes with additional keys in tests! I was thinking about unknown nodes but had no idea how could we make it fail when esquery uses its built-in fallback.